util/pyclient/bft_client.py

```diff
@@ -241,3 +247,257 @@ def get_rsi_replies(self):
         This method should be called after the send has done and before initiating a new request
         """
         return self.rsi_replies
+
+    def _process_received_msg(self, data, sender, replicas_addr, required_replies, cancel_scope):
+        """Called by child class to process a received message. At this point it's unknown if message is valid"""
+        try:
+            rsi_msg = rsi.MsgWithReplicaSpecificInfo(data, sender)
+        except (bft_msgs.MsgError , struct.error) as ex:
+            self._process_received_msg_err(ex)
```

Comment:
This method seems unnecessary to me, and breaks flow the normal linear flow control of python. The calling code can just catch the exception instead.

As it's not used in either implementation I suggest just removing it from the abstract class and inheriting classes altogether.

Reply:
@andrewjstone 
Catching the exception inside won't help since the handling is different for each type of client, so handling it must be done by the client code.
UDP client re-raise the exception inside _process_received_msg_err.
TLS client ignores it inside _process_received_msg_err.
There are some other solutions for that(like passing a flag inside MsgWithReplicaSpecificInfo), but all looked less favorable to me.

Reply:
I believe you are missing my point. This call: `_process_received_msg` is called by the concrete implementations, and so the UDP code or TLS code that calls this function can handle the exception directly and then do the same thing they would do in the callback. The callback is just an extra level of indirection. 

For example, the TLS callback swallows the exception , as you point out, but it could just as easily do that by wrapping the call to `_process_received_msg` inside `recv_data` with a `try` and a `pass` in the `except` clause.

In the UDP client case, it's even easier. You don't have to change the calling code at all since the exception is re-raised in the callback. Instead of forcing the UDP code to implement this callback and re-raise an exception, you just let the exception flow normally through the call chain by not catching it in the first place. 

Introducing this method just means that some clients have to do more work than they would otherwise, and you further complicate this method by adding the exception catch and then calling the callback.

Reply:
I'll explain my logic when writing this way:
_process_received_msg does a lot of things.
When I wrap an except around it, how can I be sure I got it in a specific place like in the current implementation?
I agree it simplifies the code, but It looks riskier to me.. 
I wanted to be sure the exception happens inside rsi.MsgWithReplicaSpecificInfo and not anywhere else. Do you suggest to check the exception location?

Reply:
Determining why the error occurred is what the exception type is for. This is similar to a return value in more procedural or functional languages. Taking your existing strategy to it's logical conclusion would require a separate callback for each line that could possibly fail.

Reply:
Ok, I've made the changes as you suggested. Thanks!